### PR TITLE
Support EIP1559 DynamicFee transactions

### PR DESCRIPTION
**Pull request by samsondav**
See the changelog for details.

This has been running on testnet for several days now without issue. I’d say its ready for another review and then merge (since its behind a default-off feature flag anyway).

### CODE REVIEW

File: core/services/eth/models.go [hunk 1]
```diff
@@ -171,6 +173,7 @@ func (h *Head) UnmarshalJSON(bs []byte) error {
 	h.Number = (*big.Int)(jsonHead.Number).Int64()
 	h.ParentHash = jsonHead.ParentHash
 	h.Timestamp = time.Unix(int64(jsonHead.Timestamp), 0).UTC()
+	h.BaseFeePerGas = (*utils.Big)((*big.Int)(jsonHead.BaseFeePerGas))
```
Comment (jmank88) [inline core/services/eth/models.go:176]: Should be able to convert directly, no?
```suggestion
	h.BaseFeePerGas = (*utils.Big)(jsonHead.BaseFeePerGas)
```

File: core/services/gas/block_history_estimator.go [hunk 2]
```diff
@@ -424,10 +536,34 @@ func (b *BlockHistoryEstimator) EffectiveGasPrice(block Block, tx Transaction) *
...
+		}
+		if block.BaseFeePerGas == nil {
+			return nil
+		}
+		effectiveTipCap := big.NewInt(0).Sub(tx.GasPrice, block.BaseFeePerGas)
+		if effectiveTipCap.Cmp(big.NewInt(0)) < 0 {
+			b.logger.Warnw("BlockHistoryEstimator: invariant violated: GasPrice - BaseFeePerGas >= 0", "block", block, "tx", tx)
+			return nil
+		}
+		return effectiveTipCap
```
Comment (jmank88) [inline core/services/gas/block_history_estimator.go:567]: What about splitting the default in to two cases?
```suggestion
	case 0x0, 0x1:
		if tx.GasPrice == nil {
			return nil
		}
		if block.BaseFeePerGas == nil {
			return nil
		}
		effectiveTipCap := big.NewInt(0).Sub(tx.GasPrice, block.BaseFeePerGas)
		if effectiveTipCap.Cmp(big.NewInt(0)) < 0 {
			b.logger.Warnw("BlockHistoryEstimator: invariant violated: GasPrice - BaseFeePerGas >= 0", "block", block, "tx", tx)
			return nil
		}
		return effectiveTipCap
	default:
		b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: ignoring unknown transaction type %v", tx.Type), "block", block, "tx", tx)
		return nil
```

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 3]
```diff
@@ -504,7 +516,8 @@ func saveAttempt(db *gorm.DB, etx *EthTx, attempt EthTxAttempt, NewAttemptState
 }
 
 func (eb *EthBroadcaster) tryAgainBumpingGas(sendError *eth.SendError, etx EthTx, attempt EthTxAttempt, initialBroadcastAt time.Time) error {
-	bumpedGasPrice, bumpedGasLimit, err := eb.estimator.BumpGas(attempt.GasPrice.ToInt(), etx.GasLimit)
+	// TODO: handle eip1559 attempts
```
Comment (jmank88) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:519]: Still todo?

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: core/services/bulletprooftxmanager/eth_broadcaster.go (hunk 3 above)
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_broadcaster.go]: Yup, fixed it now.

**Review by jmank88 — DISMISSED** [review]

File: core/store/migrate/migrations/0064_dynamic_fee_txes.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+ALTER TABLE eth_txes ADD COLUMN access_list jsonb;
+ALTER TABLE eth_tx_attempts
+	ADD COLUMN tx_type smallint NOT NULL DEFAULT 0,
+	ADD COLUMN gas_tip_cap numeric(78,0),
+	ADD COLUMN gas_fee_cap numeric(78,0),
+	ADD CONSTRAINT chk_tx_type_is_byte CHECK (
+		tx_type >= 0 AND tx_type <= 255
+	),
+	ADD CONSTRAINT chk_legacy_or_dynamic CHECK (
```
Comment (connorwstein) [inline core/store/migrate/migrations/0064_dynamic_fee_txes.sql:11]: worth it to write a little migration test that goes up to 63, adds a bunch of ethtxes/attempts and then migrates to 64 and adds a mix of legacy and dynamic ethtxes/attempts?

File: docs/CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,47 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+EIP-1559 mode is off by default but can be enabled on a per-chain basis or globally.
+
+To enable globally, set `EVM_EIP1559_DYNAMIC_FEES=true`. Set with caution, if you set this on a chain that does not actually support EIP-1559 your node will be broken.
+
+In EIP-1559 mode, the total price for the transaction is the minimum of base fee + tip cap and fee cap. More information can be found on the [official EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md).
+
+Chainlink's implementation of this is to set a large fee cap and modify the tip cap to control confirmation speed of transactions. So, when in EIP1559 mode, the tip cap takes the place of gas price roughly speaking.
+
+This should help to save gas on spikes: Chainlink will react faster on the upleg and avoid overpaying on the downleg.
```
Comment (connorwstein) [inline docs/CHANGELOG.md:22]: I think this is a little confusing - we don't save gas on the upleg right?

File: docs/CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,47 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+Chainlink now includes experimental support for submitting transactions using type 0x2 (EIP-1559) envelope.
+
+EIP-1559 mode is off by default but can be enabled on a per-chain basis or globally.
+
+To enable globally, set `EVM_EIP1559_DYNAMIC_FEES=true`. Set with caution, if you set this on a chain that does not actually support EIP-1559 your node will be broken.
+
+In EIP-1559 mode, the total price for the transaction is the minimum of base fee + tip cap and fee cap. More information can be found on the [official EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md).
+
+Chainlink's implementation of this is to set a large fee cap and modify the tip cap to control confirmation speed of transactions. So, when in EIP1559 mode, the tip cap takes the place of gas price roughly speaking.
```
Comment (connorwstein) [inline docs/CHANGELOG.md:20]: should we mention that for the most part the eip itself should help with adjusting fees to demand via the base fee? this kind of makes it seem like the eip does nothing

File: core/internal/testutils/configtest/general_config.go [hunk 7]
```diff
@@ -579,6 +589,20 @@ func (c *TestGeneralConfig) GlobalEthTxReaperThreshold() (time.Duration, bool) {
 	return c.GeneralConfig.GlobalEthTxReaperThreshold()
 }
 
+func (c *TestGeneralConfig) GlobalEvmEIP1559DynamicFees() (bool, bool) {
```
Comment (connorwstein) [inline core/internal/testutils/configtest/general_config.go:592]: I think we could get some great coverage by enabling this in a bunch of integration tests which use the geth sim, then you'd basically be testing all our features against a real geth node. Wdyt?

File: core/chains/evm/config/config.go [hunk 8]
```diff
@@ -849,6 +853,58 @@ func (c *chainScopedConfig) BalanceMonitorEnabled() bool {
...
+		return c.persistedCfg.EvmEIP1559DynamicFees.Bool
+	}
+	return c.defaultSet.eip1559DynamicFees
+}
+
+// EvmGasFeeCap is the fixed amount to set the fee cap on DynamicFee transactions
+// The recommended way to use it is to set this value to something very large
+// and control prices using the gas tip cap instead
+func (c *chainScopedConfig) EvmGasFeeCap() *big.Int {
+	return c.EvmMaxGasPriceWei()
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/chains/evm/config/config.go:875]: Are you suggesting adding a new variable to control this?

I'm not sure what that would add, since the effective meaning is identical. "This is the maximum I ever want to pay for gas" maps pretty much perfectly to the GasFeeCap.

[the remark this answers is not in the document]
Comment (samsondav) [reply core/chains/evm/config/config.go:875]: I created new env vars for all the others because they don't have a 1:1 mapping of meaning

File: core/services/bulletprooftxmanager/attempts.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+func NewDynamicFeeAttempt(cfg Config, ks KeyStore, chainID *big.Int, etx EthTx, fee gas.DynamicFee, gasLimit uint64) (attempt EthTxAttempt, err error) {
+	if err = validateDynamicFeeGas(cfg, fee, gasLimit, etx); err != nil {
+		return attempt, errors.Wrap(err, "error validating gas")
+	}
+
+	var al types.AccessList
+	if etx.AccessList.Valid {
+		al = etx.AccessList.AccessList
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/attempts.go:29]: There is no followup card, but it would be trivial to enable support for it after this PR. What job types do you think might benefit?

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 10]
```diff
@@ -250,13 +250,25 @@ func (eb *EthBroadcaster) processUnstartedEthTxs(fromAddress gethCommon.Address)
...
-		gasPrice, gasLimit, err := eb.estimator.EstimateGas(etx.EncodedPayload, etx.GasLimit)
-		if err != nil {
-			return errors.Wrap(err, "failed to estimate gas")
-		}
-		a, err := NewAttempt(eb.config, eb.ethClient, eb.keystore, eb.chainID, *etx, gasPrice, gasLimit)
-		if err != nil {
-			return errors.Wrap(err, "processUnstartedEthTxs failed")
+		var a EthTxAttempt
+		if eb.config.EvmEIP1559DynamicFees() {
+			fee, gasLimit, err := eb.estimator.GetDynamicFee(etx.GasLimit)
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_broadcaster.go:255]: It does not bump the gas limit normally. The reason we pass the gaslimit through the estimator is twofold:

1. In Optimismv1 mode, gas limit changes depending on gas price
2. Multipliers/modifiers can be applied to the gas limit right before sending

File: docs/CHANGELOG.md (hunk 5 above)
Comment (samsondav) [reply docs/CHANGELOG.md:22]: It's actually quite difficult to predict. We might save some gas on the upleg too because we respond to tip, not the total, so as the base fee increases we dont unnecessarily boost our payment.

In the end, the only way to know will be to test it out in prod.

File: core/internal/testutils/configtest/general_config.go (hunk 7 above)
Comment (samsondav) [reply core/internal/testutils/configtest/general_config.go:592]: Yup I agree. Which integration tests do you have in mind?

File: core/services/bulletprooftxmanager/eth_broadcaster_test.go [hunk 11]
```diff
@@ -227,6 +232,71 @@ func TestEthBroadcaster_ProcessUnstartedEthTxs_Success(t *testing.T) {
...
+		require.NoError(t, db.Save(&eipTxWithoutAl).Error)
+
+		// Do the thing
+		require.NoError(t, eb.ProcessUnstartedEthTxs(keyState))
+
+		// Check eipTxWithAl and it's attempt
+		// This was the earlier one sent so it has the lower nonce
+		eipTxWithAl, err := cltest.FindEthTxWithAttempts(db, eipTxWithAl.ID)
+		require.NoError(t, err)
+		assert.False(t, eipTxWithAl.Error.Valid)
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_broadcaster_test.go:278]: I tried to do this and ran into numerous annoyances, for example comparing BroadcastAt doesn't work (firstly because we dont care what it is as long as its not nil, secondly because timestamps lose precision when persisted). Other ones are things like comparing pointers vs. the exact values etc, in the end it didn't add much.

File: core/store/migrate/migrations/0064_dynamic_fee_txes.sql (hunk 4 above)
Comment (samsondav) [reply core/store/migrate/migrations/0064_dynamic_fee_txes.sql:11]: We don't have migration tests any more, and I think not having them is the right way to go.

With that having been said, I have run some manual tests on a real database and it works ok.

File: docs/CHANGELOG.md (hunk 6 above)
Comment (samsondav) [reply docs/CHANGELOG.md:20]: I have tried to rephrase this to make it clearer but it's quite hard to explain and really just a guess; we need to test in practice.

File: core/chains/evm/config/config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply core/chains/evm/config/config.go:873]: yeah thats what I was suggesting. I could see a nop thinking "I guess this MAX_GAS_PRICE doesn't have any effect since there are no gas "prices" with 1559". I think as long as its documented clearly its fine as is

File: core/services/bulletprooftxmanager/eth_broadcaster.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply core/services/bulletprooftxmanager/eth_broadcaster.go:251]: thanks for clarifying

File: core/store/migrate/migrations/0064_dynamic_fee_txes.sql (hunk 4 above)
Comment (connorwstein) [reply core/store/migrate/migrations/0064_dynamic_fee_txes.sql]: I think it definitely makes sense to not run the migration tests in the CI, but we still want to have them for documentation and reproducibility of manual tests IMO

File: core/internal/testutils/configtest/general_config.go (hunk 7 above)
Comment (connorwstein) [reply core/internal/testutils/configtest/general_config.go:592]: I'd say TestIntegration_OCR, TestIntegration_DirectRequest,  TestFluxMonitor_Deviation, TestIntegration_VRF_JPV2 and TestKeeperEthIntegration

File: core/services/bulletprooftxmanager/attempts.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+	attempt.ChainSpecificGasLimit = gasLimit
+	attempt.TxType = 2
+	return attempt, nil
+}
+
+var Max256BitUInt = big.NewInt(0).Exp(big.NewInt(2), big.NewInt(256), nil)
+
+// validateDynamicFeeGas is a sanity check - we have other checks elsewhere, but this
+// makes sure we _never_ create an invalid attempt
+func validateDynamicFeeGas(cfg Config, fee gas.DynamicFee, gasLimit uint64, etx EthTx) error {
```
Comment (jmank88) [inline core/services/bulletprooftxmanager/attempts.go:51]: Probably worth locking down with some tests for invalid cases?

File: core/services/gas/block_history_estimator.go [hunk 13]
```diff
@@ -424,10 +536,33 @@ func (b *BlockHistoryEstimator) EffectiveGasPrice(block Block, tx Transaction) *
...
+		b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: ignoring unknown transaction type %v", tx.Type), "block", block, "tx", tx)
+		return nil
+	}
+}
+
+func (b *BlockHistoryEstimator) EffectiveTipCap(block Block, tx Transaction) *big.Int {
+	switch tx.Type {
+	case 0x2:
+		return tx.MaxPriorityFeePerGas
+	case 0x0, 0x1:
```
Comment (jmank88) [inline core/services/gas/block_history_estimator.go:551]: Should this case be limited to `0x0`, like  in `bumpGas()`?

Would it be helpful to define some constants for these types? Might be more readable in some places, but repetitive in others. Maybe better in the future with additional types. :shrug:

**Review by jmank88 — CHANGES_REQUESTED** [review]

File: core/services/bulletprooftxmanager/attempts.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply core/services/bulletprooftxmanager/attempts.go:22]: I know that VRF jobs would because of how the contract is set up. I suspect basically all jobs could benefit, but how exactly I'd have to read the contract code more carefully. Probably needs a spike first

[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/attempts.go:22]: It's basically trivial to specify an AccessList at insert-time. We'd have to take it on a job-by-job basis

[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/attempts.go:22]: Created this story to track: https://app.shortcut.com/chainlinklabs/story/17478/use-accesslists-where-appropriate-to-save-gas

File: core/chains/evm/config/config.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/chains/evm/config/config.go:873]: We definitely need to update the official docs once this is released.

File: core/services/gas/block_history_estimator.go (hunk 13 above)
Comment (samsondav) [reply core/services/gas/block_history_estimator.go:557]: It should not be limited like `bumpGas` because bumpGas controls transactions that we send, and we never send `0x1`.

However, `0x1` transactions _can_ appear on-chain and we should handle those correctly when using the tx for estimation.

I thought about defining constants but it seems unwieldy when its really simple to just type `0x0` or `0x2` etc and everyone that's familiar with Ethereum will know what that means. Seems to be more clear just to be explicit.

File: core/internal/testutils/configtest/general_config.go (hunk 7 above)
Comment (samsondav) [reply core/internal/testutils/configtest/general_config.go:590]: Good call, I realised this isn't compatible with Keeper yet.

**Review by jmank88 — APPROVED** [review]
